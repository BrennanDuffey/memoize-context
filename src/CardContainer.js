import React, { Component } from 'react';
import QuestionCard from './QuestionCard.js';
import AnswerCard from './AnswerCard.js';

class CardContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      cardCounter: 0,
      showAnswer: false,
      isCorrect: false,
      correctAnswers: [],
      filteredQuestions: this.props.questions
    }
  };

  updateIsCorrect = (boolean) => {
    this.setState({
      isCorrect: boolean
    });
  };

  updateFilteredQuestions = () => {
    let filteredQuestions = this.state.filteredQuestions
      .filter(question => JSON.parse(localStorage.getItem('correct answers').includes(question)))
    this.setState({filteredQuestions: filteredQuestions}, () => {console.log(this.state.filteredQuestions)})
  }

  updateCorrectAnswers = () => {
    let correctAnswers = this.state.correctAnswers;
    if(this.state.isCorrect) {
      correctAnswers.push(this.props.questions[this.state.cardCounter-1])
    }
    this.setState({correctAnswers: correctAnswers}, () => {
      localStorage.setItem('correct answers', JSON.stringify(this.state.correctAnswers))
    });
  }

  toggleShowAnswer = () => {
    this.setState({
      showAnswer: !this.state.showAnswer
    });
  };

  updateCounter = () => {
    let newCounter = this.state.cardCounter + 1;
    this.setState({cardCounter: newCounter});
  };

  nextQuestion = () => {
    this.updateCorrectAnswers();
    this.updateCounter();
    this.toggleShowAnswer();
  };


  render () {
    let response;
    if (this.state.isCorrect) {
      response = 'https://media.giphy.com/media/PS7d4tm1Hq6Sk/giphy.gif';
    } else {
      response = 'https://media.giphy.com/media/1455m6M8jFgCE8/giphy.gif';
    }
    if(!this.state.cardCounter) {
      return (
        <section>
          <h2>Hello and Welcome to Context Quiz please press the start button to begin</h2>
          <button onClick={this.updateCounter}> Start </button>
        </section>
      )
    } else if (this.state.showAnswer) {
      return (
        <section>
          <img src={response} />
          <AnswerCard {...this.props.questions[this.state.cardCounter-1]}/>
          <button onClick={this.nextQuestion}> Next </button>
        </section>
      )
    } else if (this.state.cardCounter === 32) {
      return (
        <section>
          <h2>Quiz completed!</h2>
        </section>
      )
    } else {
      return (
        <section>
          <QuestionCard {...this.props.questions[this.state.cardCounter-1]} 
                        toggleShowAnswer={this.toggleShowAnswer} 
                        updateIsCorrect={this.updateIsCorrect} 
                        updateCorrectAnswers={this.updateCorrectAnswers}/>
        </section>
      )
    }
  }
}
export default CardContainer;