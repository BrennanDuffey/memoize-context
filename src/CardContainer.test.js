import React from 'react';
import CardContainer from './CardContainer.js';
import { shallow } from 'enzyme';

let mockQuestions = [
  {
    "question": "When the context of this is described what is context actually?",
    "answers": ["The scope that this is invoked in.", "Whatever this is bound to.", "The object that is executing the javascript code."],
    "correctAnswer": "The object that is executing the javascript code.",
    "codeSnippet": null,
    "id": 1
  },
  {
    "question": "When is the value of this set in an ES5 function?",
    "answers": ["When the function is created.", "When the function is executed."],
    "correctAnswer": "When the function is executed.",
    "codeSnippet": null,
    "id": 2
  }
];
const mockGetLocalStorage = jest.fn();


describe('Answer components', () => {
  
  let wrapper;

  beforeEach(() => {
    wrapper = shallow (
      <CardContainer  
        questions={mockQuestions}
        getLocalStorage={mockGetLocalStorage}
      />
    );
  });

  it('Should render correctly with all the information passed', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should have default states', () => {
    expect(wrapper.state()).toEqual({
      cardCounter: 0,
      showAnswer: false,
      isCorrect: false,
      incorrectAnswers: []
    });
  });

  it('Should be able to update is correct', () => {
    expect(wrapper.state().isCorrect).toEqual(false);
    wrapper.instance().updateIsCorrect(true);
    expect(wrapper.state().isCorrect).toEqual(true);
  });

  it('Should be able to toggle the answer', () => {
    expect(wrapper.state().showAnswer).toEqual(false);
    wrapper.instance().toggleShowAnswer();
    expect(wrapper.state().showAnswer).toEqual(true);
  });

  it('Should be able to reset the counter and incorrectAnswers', () => {
    wrapper.state().cardCounter = 1;
    wrapper.state().incorrectAnswers = [1, 2, 3]
    wrapper.instance().resetCounter();
    expect(wrapper.state().cardCounter).toEqual(0);
    expect(wrapper.state().incorrectAnswers).toEqual([]);
  });

  it('Should be able to updateCounter', () => {
    expect(wrapper.state().cardCounter).toEqual(0);
    wrapper.instance().updateCounter();
    expect(wrapper.state().cardCounter).toEqual(1);
    wrapper.instance().updateCounter();
    expect(wrapper.state().cardCounter).toEqual(2);
  });

  it('Should be able to update correct Answers when someone is wrong', () => {
    wrapper.state().cardCounter = 1;
    expect(wrapper.state().incorrectAnswers).toEqual([]);
    wrapper.instance().updateCorrectAnswers();
    expect(wrapper.state().incorrectAnswers).toEqual([
      {
        "question": "When the context of this is described what is context actually?",
        "answers": ["The scope that this is invoked in.", "Whatever this is bound to.", "The object that is executing the javascript code."],
        "correctAnswer": "The object that is executing the javascript code.",
        "codeSnippet": null,
        "id": 1
      }
    ]);
  });

  it('Should call a bunch of methods to be called from nextQuestion', () => {
    expect(wrapper.state().cardCounter).toEqual(0);
    wrapper.instance().nextQuestion()
    expect(wrapper.state().cardCounter).toEqual(1);
  });

  describe('Conditional rendering', () => {
    
    it('Should render based off cardCounter > 0', () => {
      wrapper.state().cardCounter = 1;
      expect(wrapper).toMatchSnapshot();
    });

    it('Should render based off cardCounter = questions.length + 1', () => {
      wrapper.state().cardCounter = 3;
      expect(wrapper).toMatchSnapshot();
    });

    it('Should render differently if state of show answer is true', () => {
      wrapper.state().showAnswer = true;
      expect(wrapper).toMatchSnapshot();
    });
  });
});