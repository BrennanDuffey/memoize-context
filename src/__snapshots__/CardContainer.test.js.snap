// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer components Conditional rendering Should render based off cardCounter = questions.length + 1 1`] = `
<section
  className="flash-card"
>
  <h1>
    Hello and Welcome to Context Quiz please press the start button to begin
  </h1>
  <button
    onClick={[Function]}
  >
     Start 
  </button>
</section>
`;

exports[`Answer components Conditional rendering Should render based off cardCounter > 0 1`] = `
<section
  className="flash-card"
>
  <h1>
    Hello and Welcome to Context Quiz please press the start button to begin
  </h1>
  <button
    onClick={[Function]}
  >
     Start 
  </button>
</section>
`;

exports[`Answer components Conditional rendering Should render differently if state of show answer is true 1`] = `
<section
  className="flash-card"
>
  <h1>
    Hello and Welcome to Context Quiz please press the start button to begin
  </h1>
  <button
    onClick={[Function]}
  >
     Start 
  </button>
</section>
`;

exports[`Answer components Should render correctly with all the information passed 1`] = `
<section
  className="flash-card"
>
  <h1>
    Hello and Welcome to Context Quiz please press the start button to begin
  </h1>
  <button
    onClick={[Function]}
  >
     Start 
  </button>
</section>
`;
